src/uct/ib/base/ib_md.c

```diff
@@ -1694,33 +1705,79 @@ ucs_status_t uct_ib_md_open_common(uct_ib_md_t *md,
 
     return UCS_OK;
 
-err_dealloc_pd:
-    ibv_dealloc_pd(md->pd);
 err_cleanup_device:
     uct_ib_device_cleanup(&md->dev);
 err_release_stats:
     UCS_STATS_NODE_FREE(md->stats);
+err_device_config_release:
+    uct_ib_md_release_device_config(md);
 err:
     return status;
 }
 
-void uct_ib_md_close(uct_md_h uct_md)
+void uct_ib_md_close_common(uct_ib_md_t *md)
 {
-    uct_ib_md_t *md = ucs_derived_of(uct_md, uct_ib_md_t);
-
-    /* Must be done before md->ops->cleanup, since it can call functions from
-     * md->ops */
     uct_ib_md_release_reg_method(md);
-    md->ops->cleanup(md);
-    uct_ib_md_release_device_config(md);
-    uct_ib_device_cleanup_ah_cached(&md->dev);
-    ibv_dealloc_pd(md->pd);
     uct_ib_device_cleanup(&md->dev);
-    ibv_close_device(md->dev.ibv_context);
     UCS_STATS_NODE_FREE(md->stats);
+    uct_ib_md_release_device_config(md);
+}
+
+void uct_ib_md_device_context_close(struct ibv_context *ctx)
+{
+    int ret = ibv_close_device(ctx);
+    if (ret != 0) {
+        ucs_warn("ibv_close_device() failed: %m");
+    }
+}
+
+uct_ib_md_t* uct_ib_md_alloc(size_t size, const char *name,
```

Comment:
if we manage to move ibv_alloc_pd to open_common then this function would be redundant

Reply:
unfortunately, we can't move `ibv_alloc_pd` to `open_common`